File: exercises/C/utils/read_options.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "read_options.h"
  
void read_options(int argc, char**argv, 
                  int* gridSizeX,
                  int* gridSizeY,
                  int* iter,
                  float* xMin,
                  float* xMax,
                  float* yMin,
                  float* yMax)
{
  
    int count;
    int tmp;
    float ftmp;
  
    /* loop over command line arguments */
    while ( (count = getopt(argc, argv, "s:S:i:x:X:y:Y:mjh")) != -1 ) {
        switch (count) {
      
        case 'S':
            /* option -S: grid size in X dimension */
            tmp = atoi(optarg);
            if ( tmp >= 4 ) *gridSizeX = tmp;
            else fprintf(stderr, "Using default grid size.\n");
            break;
        case 'i':
            /* option -i: number of iterations for the calculation */
            tmp = atoi(optarg);
            if ( tmp > 0 ) *iter = tmp;
            else fprintf(stderr, "Using default number of iterations.\n");
            break;
      
        case 'x':
            /* option -x:
               for Mandelbrot set: xmin coordinate of the image */
            ftmp = atof(optarg);
            *xMin = ftmp;
            break;

        case 'X':
            /* option -X: xmax coordinate of the image */
            ftmp = atof(optarg);
            *xMax = ftmp;
            break;
      
        case 'y':
            /* option -y:
               for Mandelbrot set: ymin coordinate of the image
               for Julia set: y starting point coordinate
            */
            ftmp = atof(optarg);
            *yMin = ftmp;
            break;
      
        case 'Y':
            /* option -Y: ymax coordinate of the image */
            ftmp = atof(optarg);
            *yMax = ftmp;
            break;

        case '?':
        case 'h':
            printf("Usage:\n");
            printf("mandelbrot [-SixXyYh]\n");
            printf("   -S NPIXEL    Set number of pixels in X dimension of image\n");
            printf("                Y dimension is scaled to keep square aspect ratio\n");
            printf("   -i ITS       Set max number of iterations for a point to be inside\n");
            printf("   -x XMIN      Set xmin coordinate\n");
            printf("   -X XMAX      Set xmax coordinate\n");
            printf("   -y YMIN      Set ymin coordinate (mandelbrot) or starting point (Julia)\n");
            printf("   -Y YMAX      Set ymax coordinate\n");
            printf("   -h           Show this help\n");
            exit(-1);
        }
    

    } 
  
    
    /* check that xMin is smaller than xMax, if not, reassign the default values */
    if ( *xMin >= *xMax ) {
        fprintf(stderr, "WARNING: xMin has to be smaller than xMax.\n Reassign default values.\n");
        *xMin = XMIN;
        *xMax = XMAX;
    }
  
    /* check that yMin is smaller than yMax, if not, reassign the default values */
    if ( *yMin >= *yMax ) {
        fprintf(stderr, "WARNING: yMin has to be smaller than yMax.\n Reassign default values.\n");
        *yMin = YMIN;
        *yMax = YMAX;
    }

    *gridSizeY = (int)(*gridSizeX * (*yMax - *yMin)/(*xMax - *xMin));
}
 

File: exercises/F/utils/read_options.f90
module read_opt
  implicit none
  complex, parameter :: MIN = (-2.0, -1.2)
  complex, parameter :: MAX = (1, 1.2)

  integer, parameter :: GRIDSIZE_X = 768

  integer, parameter :: ITERATIONS = 5000
contains

  subroutine read_options(gridsizex, gridsizey, iter, cmin, cmax)
    integer, intent(out) :: gridsizex
    integer, intent(out) :: gridsizey
    integer, intent(out) :: iter
    complex, intent(out) :: cmin
    complex, intent(out) :: cmax

    integer :: i
    integer :: nargs
    real :: tmp
    character(len=128) :: arg
    character(len=128) :: opt
    nargs = command_argument_count()
    i = 1

    iter = ITERATIONS
    gridsizex = GRIDSIZE_X
    cmin = MIN
    cmax = MAX

    do
       if ( i .gt. nargs ) exit
       call get_command_argument(i, value=arg)
       if ( i+1 .gt. nargs .and. (trim(arg(2:)) .ne. 'h') ) then
          write(*,*)"Argument `", trim(arg), "' needs a value"
          arg = '?'
       else
          call get_command_argument(i+1, value=opt)
       end if

       select case(trim(arg(2:)))
       case ('S')
          read(opt, *)gridsizex
          if ( gridsizex < 4 ) gridsizex = GRIDSIZE_X
       case ('i')
          read(opt, *)iter
       case ('x')
          read(opt, *)tmp
          cmin = cmplx(tmp, aimag(cmin))
       case ('X')
          read(opt, *)tmp
          cmax = cmplx(tmp, aimag(cmax))
       case ('y')
          read(opt, *)tmp
          cmin = cmplx(real(cmin), tmp)
       case ('Y')
          read(opt, *)tmp
          cmax = cmplx(real(cmax), tmp)
       case default
          write(*, *)'Usage:'
          write(*, *)'mandelbrot [-SixXyY]'
          write(*, *)'   -S NPIXEL    Set number of pixels in X dimension of image'
          write(*, *)'                Y dimension is scaled to ensure pixels are square'
          write(*, *)'   -i ITS       Set max number of iterations for a point to be inside'
          write(*, *)'   -x XMIN      Set xmin coordinate'
          write(*, *)'   -X XMAX      Set xmax coordinate'
          write(*, *)'   -y YMIN      Set ymin coordinate'
          write(*, *)'   -Y YMAX      Set ymax coordinate'
          write(*, *)'   -h           Show this help'
          stop
       end select
       i = i + 2
    end do

    if ( real(cmin) .ge. real(cmax) ) cmin = cmplx(real(MIN), aimag(cmin))
    if ( aimag(cmin) .ge. aimag(cmax) ) cmin = cmplx(real(cmin), aimag(MIN))
    gridsizey = int(gridsizex * aimag(cmax - cmin) / real(cmax - cmin))

  end subroutine read_options
end module read_opt
